src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -218,7 +218,73 @@ public virtual int Next(int maxValue)
             return (int)(Sample() * maxValue);
         }
 
-        /*=====================================Next=====================================
+        private long FullLong()
+        {
+            long result = 0;
+
+            for (int i = 0; i < 4; i++)
+            {
+                short part;
+                unchecked
+                {
+                    part = (short) InternalSample();
+                }
+
+                result |= (long) part;
+                result <<= 16;
+            }
+
+            return result;
+        }
+
+        /*==================================NextInt64===================================
+        **Returns: An long [0..long.maxValue)
+        **Arguments: maxValue -- One more than the greatest legal return value.
+        **Exceptions: None.
+        ==============================================================================*/
+        public virtual long NextInt64()
+        {
+            return FullLong();
+        }
+
+        /*==================================NextInt64===================================
+        **Returns: A long [0..maxValue)
+        **Arguments: maxValue -- One more than the greatest legal return value.
+        **Exceptions: None.
+        ==============================================================================*/
+        public virtual long NextInt64(long maxValue)
+        {
+            if (maxValue < 0)
+                throw new ArgumentOutOfRangeException(nameof(maxValue), SR.Format(SR.ArgumentOutOfRange_MustBePositive, nameof(maxValue)));
+
+            if (maxValue == 0)
+                return 0;
+
+            long fullLong = (long) FullLong();
+            return (long.MaxValue & fullLong) % maxValue;
+        }
+
+        /*==================================NextInt64===================================
+        **Returns: A long [minvalue..maxvalue)
+        **Arguments: minValue -- the least legal value for the Random number.
+        **           maxValue -- One greater than the greatest legal return value.
+        **Exceptions: None.
+        ==============================================================================*/
+        public virtual long NextInt64(long minValue, long maxValue)
+        {
+            if (minValue > maxValue)
+            {
+                throw new ArgumentOutOfRangeException(nameof(minValue), SR.Format(SR.Argument_MinMaxValue, nameof(minValue), nameof(maxValue)));
+            }
+
+            long range = maxValue - minValue;
+            long range1 = range / 2;
+            long range2 = range - range1;
+
+            return NextInt64(range1 + 1) + NextInt64(range2) + minValue;
```

Comment:
This is attempting to dodge additional overflow issues, but I'm unsure whether this will properly distribute values over the entire range.  Possibly suspect.